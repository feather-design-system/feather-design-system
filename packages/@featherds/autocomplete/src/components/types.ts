import { Ref, PropType } from "vue";
import HighlightProps from "./Highlight/HighlightProps";
import { InputWrapperProps, InputSubTextProps } from "@featherds/input-helper";
interface IAutocompleteItem extends IAutocompleteItemType {
  _new?: string;
  _text: string;
  _pre?: IAutocompleteChipIcon;
}
interface IAutocompleteItemType {
  [k: string]: unknown;
}

interface IAutocompleteGridColumn {
  prop: keyof IAutocompleteItemType;
  title: string;
  align?: "right" | "left" | "center";
}

interface IAutocompleteResultRender {
  reset: () => void;
  handleKeyPress: (
    e: KeyboardEvent,
    internalResults: IAutocompleteItemType[]
  ) => boolean;
  active: {
    row: number;
    col?: number;
  };
  first: () => void;
}

export enum AutocompleteTypes {
  multi = "multi",
  single = "single",
}
interface IStrategyOptions {
  selectionLimit: Ref<number>;
  selectionLimitReached: Ref<boolean>;
  modelValue: Ref<IAutocompleteItemType | IAutocompleteItemType[]>;
  textProp: Ref<string>;
  input: Ref<HTMLInputElement>;
  query: Ref<string>;
  emitSearch: () => void;
  forceCloseResults: Ref<boolean>;
  allowNew: Ref<boolean>;
  internalResults: Ref<IAutocompleteItemType[]>;
}

interface IAutocompleteChipIcon {
  icon: unknown;
  title: string;
}

export type {
  IStrategyOptions,
  IAutocompleteGridColumn,
  IAutocompleteItemType,
  IAutocompleteItem,
  IAutocompleteChipIcon,
  IAutocompleteResultRender,
};

export const LABELS = {
  noResults: "No results found",
  minChar: "Enter ${min} characters to search",
  clear: "Clear value",
  selectionLimit: "Selection limit reached",
  new: "New",
  remove: "Remove",
};
export const props = {
  id: {
    type: String,
  },
  modelValue: {
    type: [Array, Object] as PropType<
      IAutocompleteItemType[] | IAutocompleteItemType
    >,
  },
  results: {
    type: Array as PropType<IAutocompleteItemType[]>,
    default: () => [],
  },
  textProp: {
    type: String as unknown as PropType<keyof IAutocompleteItemType>,
    default: "_text",
  },
  loading: {
    type: Boolean,
    default: false,
  },
  minChar: {
    type: Number,
    default: 0,
  },
  allowNew: {
    type: Boolean,
    default: false,
  },
  selectionLimit: {
    type: Number,
  },
  newMatcher: {
    type: Function,
    default: (
      item: IAutocompleteItemType,
      query: string,
      comp: { textProp: string }
    ) => {
      return (
        (item[comp.textProp] as string).toString().toLowerCase() ===
        query.toLowerCase()
      );
    },
  },
  type: {
    type: String as PropType<keyof typeof AutocompleteTypes>,
    required: true,
    validator: (v: keyof typeof AutocompleteTypes) => {
      // The value must match either
      return !!AutocompleteTypes[v];
    },
  },
  labels: {
    type: Object as PropType<Partial<typeof LABELS>>,
    default() {
      return LABELS;
    },
  },
  gridConfig: {
    type: Array as PropType<IAutocompleteGridColumn[]>,
  },
  schema: {
    type: Object,
    required: false,
  },
  ...HighlightProps,
  ...InputSubTextProps,
  ...InputWrapperProps,
} as const;

export const emits = {
  "update:modelValue": (
    value: IAutocompleteItemType | IAutocompleteItemType[] | undefined
  ) => true,
  new: (value: string) => true,
  search: (value: string) => true,
};
