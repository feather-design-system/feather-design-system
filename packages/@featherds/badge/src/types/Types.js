const TYPES = {
  INFO: "info",
  ERROR: "error",
};

export default TYPES;
