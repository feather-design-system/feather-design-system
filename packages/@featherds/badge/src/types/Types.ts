export enum BadgeTypes {
  info = "info",
  error = "error",
}

export interface IBadge {
  type: BadgeTypes;
  title: string;
}
