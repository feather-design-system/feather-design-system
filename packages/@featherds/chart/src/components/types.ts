export type ShirtSize = "xs" | "sm" | "md" | "lg" | "xl";

export type FeatherChartType = "bar";

export type FeatherBarChartData = {
  labels: Array<string>;
  datasets: Array<{
    label: string;
    data: Array<string | number>;
    backgroundColor?: string;
  }>;
};

export type FeatherBarChartOptions = {
  responsive?: boolean;
  maintainAspectRation?: boolean;
};

export type FeatherD3Dimensions = {
  height: number;
  width: number;
  margin: {
    top: number;
    right: number;
    bottom: number;
    left: number;
  }
};

export type FeatherD3Options = {
  units: string;
  class?: "color1" | "color2";
};


// Data can be any array of objects returned from an API or local JSON
export type FeatherD3Data = Array<unknown>;

export interface ICoordinateAxes {
  x: string,
  y: string
}
