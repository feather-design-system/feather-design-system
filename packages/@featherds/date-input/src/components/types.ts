export interface IDateDisabledConfig extends IDateRange {
  dates?: (Date | number)[];
  ranges?: IDateRange[];
  days?: number[];
  daysOfMonth?: number[];
  customPredictor?: (date: Date | number) => boolean;
}
interface IDateRange {
  to?: Date | number;
  from?: Date | number;
}
export const LABELS = {
  day: "Day",
  month: "Month",
  year: "Year",
  prevMonth: "Previous month",
  nextMonth: "Next month",
  selectMonth: "Select month",
  selectYear: "Select year",
  inCalendar: "In calendar",
  calendar: "Click for calendar",
  clear: "Clear date",
  menu: "Use arrow keys to navigate dates. Page down and page up will navigate by month. Shift page down and shift page up will navigate by year. Press escape to exit the calendar.",
};
