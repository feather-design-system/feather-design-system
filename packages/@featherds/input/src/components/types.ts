import { InputWrapperProps, InputSubTextProps } from "@featherds/input-helper";
import { PropType } from "vue";
export const props = {
  type: {
    type: String,
    default: "text",
  },
  modelValue: {
    type: [String, Number] as PropType<string | number>,
  },
  maxlength: {
    type: Number,
    required: false,
    default: 0,
  },
  schema: {
    type: Object as PropType<Record<string, any>>,
    required: false,
  },
  id: {
    type: String,
    required: false,
  },
  ...InputWrapperProps,
  ...InputSubTextProps,
} as const;

export const emits = {
  "update:modelValue": (value: string | undefined | number) => true,
};
