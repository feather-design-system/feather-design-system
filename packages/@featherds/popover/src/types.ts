// Note: introduced new types
export type Tooltip = {
  height: number;
  width: number;
};

export type Trigger = {
  height: number;
  width: number;
  left: number;
  top: number;
};

export enum PopoverPlacement {
  top = "top",
  bottom = "bottom",
  left = "left",
  right = "right",
}

export enum PointerAlignment {
  center = "center",
  left = "left",
  right = "right",
}
