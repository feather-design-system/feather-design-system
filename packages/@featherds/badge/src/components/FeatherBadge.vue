<template>
  <span class="badge-container" :class="{ inline: inline }">
    <slot />
    <div class="badges">
      <Badge
        v-for="item in orderBadges"
        :key="item.type"
        :type="item.type"
        :title="item.title"
        :siblings="orderBadges.length > 1"
      />
    </div>
  </span>
</template>
<script>
import TYPES from "../types/Types";
const BADGES = [TYPES.INFO, TYPES.ERROR];
import Badge from "./Badge";
export default {
  props: {
    badges: {
      type: [Array, Object],
      required: true,
    },
    inline: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    orderBadges() {
      const badges = Array.isArray(this.badges) ? this.badges : [this.badges];
      return badges
        .filter((x) => BADGES.indexOf(x.type.toLowerCase()) > -1)
        .sort((a, b) => {
          const aIndex = BADGES.indexOf(a.type.toLowerCase());
          const bIndex = BADGES.indexOf(b.type.toLowerCase());
          return aIndex - bIndex;
        });
    },
  },
  components: {
    Badge,
  },
};
</script>

<style lang="scss" scoped>
.badge-container {
  display: inline-block;
  position: relative;
  &.inline {
    display: inline-flex;
    align-items: center;

    .badges {
      position: static;
      margin-left: 4px;
    }
  }
  .badges {
    position: absolute;
    left: calc(100% - 5px);
    bottom: calc(100% - 5px);
    display: flex;
  }
}
</style>
