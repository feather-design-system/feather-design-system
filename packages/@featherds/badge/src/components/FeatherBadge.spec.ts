import FeatherBadge from "./FeatherBadge.vue";
import { BadgeTypes } from "../types/Types";
import { shallowMount } from "@vue/test-utils";
import { expect, describe, it } from "vitest";

const getWrapper = function (options: Record<string, unknown>) {
  return shallowMount(FeatherBadge, options);
};

describe("FeatherBadge.vue", () => {
  it("should render single icon ", () => {
    const badge = {
      type: BadgeTypes.info,
      title: "MESSAGE",
    };
    const wrapper = getWrapper({
      props: {
        badges: badge,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should render multiple icon in correct order", () => {
    const badges = [
      {
        type: BadgeTypes.error,
        title: "MESSAGE",
      },
      {
        type: BadgeTypes.info,
        title: "MESSAGE",
      },
    ];
    const wrapper = getWrapper({
      props: {
        badges: badges,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should render single icon inline", () => {
    const badge = {
      type: BadgeTypes.info,
      title: "MESSAGE",
    };
    const wrapper = getWrapper({
      props: {
        badges: badge,
        inline: true,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should render multiple icon inline in correct order ", () => {
    const badges = [
      {
        type: BadgeTypes.error,
        title: "MESSAGE",
      },
      {
        type: BadgeTypes.info,
        title: "MESSAGE",
      },
    ];
    const wrapper = getWrapper({
      props: {
        badges: badges,
        inline: true,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});
