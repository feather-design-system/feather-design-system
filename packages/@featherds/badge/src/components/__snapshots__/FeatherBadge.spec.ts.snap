// Vitest Snapshot v1

exports[`FeatherBadge.vue > should render multiple icon in correct order 1`] = `
<span
  class="badge-container"
>
  
  
  <div
    class="badges"
  >
    
    <badge-stub
      siblings="true"
      title="MESSAGE"
      type="info"
    />
    <badge-stub
      siblings="true"
      title="MESSAGE"
      type="error"
    />
    
  </div>
</span>
`;

exports[`FeatherBadge.vue > should render multiple icon inline in correct order  1`] = `
<span
  class="badge-container inline"
>
  
  
  <div
    class="badges"
  >
    
    <badge-stub
      siblings="true"
      title="MESSAGE"
      type="info"
    />
    <badge-stub
      siblings="true"
      title="MESSAGE"
      type="error"
    />
    
  </div>
</span>
`;

exports[`FeatherBadge.vue > should render single icon  1`] = `
<span
  class="badge-container"
>
  
  
  <div
    class="badges"
  >
    
    <badge-stub
      siblings="false"
      title="MESSAGE"
      type="info"
    />
    
  </div>
</span>
`;

exports[`FeatherBadge.vue > should render single icon inline 1`] = `
<span
  class="badge-container inline"
>
  
  
  <div
    class="badges"
  >
    
    <badge-stub
      siblings="false"
      title="MESSAGE"
      type="info"
    />
    
  </div>
</span>
`;
